pkg/topology/pslice/pslice.go

```diff
@@ -16,36 +16,66 @@ import (
 // in order to reduce duplicate PO calculation which is normally known and already needed in the
 // calling context.
 type PSlice struct {
-	peers     []swarm.Address // the slice of peers
-	bins      []uint          // the indexes of every proximity order in the peers slice, index is po, value is index of peers slice
+	peers     [][]swarm.Address // the slice of peers
 	baseBytes []byte
 	sync.RWMutex
+	maxBins int
 }
 
-// New creates a new PSlice.
+// New creates a new PSlice
 func New(maxBins int, base swarm.Address) *PSlice {
 	return &PSlice{
-		peers:     make([]swarm.Address, 0),
-		bins:      make([]uint, maxBins),
+		peers:     make([][]swarm.Address, maxBins),
 		baseBytes: base.Bytes(),
+		maxBins:   maxBins,
+	}
+}
+
+// Add a peer at a certain PO.
+func (s *PSlice) Add(addrs ...swarm.Address) {
+	s.Lock()
+	defer s.Unlock()
+
+	pos := make([]uint8, 0, len(addrs))
+	poCount := make([]int, s.maxBins)
+
+	for _, addr := range addrs {
+		po := s.po(addr.Bytes())
+		pos = append(pos, po)
+		poCount[po]++
+	}
+
+	for i, count := range poCount {
+		peers := s.peers[i]
+		if count > 0 && cap(peers) < len(peers)+count {
+			newPeers := make([]swarm.Address, len(peers), len(peers)+count)
+			copy(newPeers, peers)
+			s.peers[i] = newPeers
+		}
+	}
+
+	for i, addr := range addrs {
+		po := pos[i]
+
+		if e, _ := s.index(addr, po); e {
+			continue
+		}
+
+		s.peers[po] = append(s.peers[po], addr)
 	}
 }
 
 // iterates over all peers from deepest bin to shallowest.
 func (s *PSlice) EachBin(pf topology.EachPeerFunc) error {
-	s.RLock()
-	peers, bins := s.peers, s.bins
-	s.RUnlock()
 
-	if len(peers) == 0 {
-		return nil
-	}
+	for i := s.maxBins - 1; i >= 0; i-- {
 
-	var binEnd = uint(len(peers))
+		s.RLock()
```

Comment:
@acud This seems the only way to prevent data races, unless I lock the whole function at the beginning.